File: solved/isValid.ts
// https://leetcode.com/problems/valid-parentheses/
export const isValid = (s: string): boolean => {
  if (s.length % 2 !== 0) return false

  const map = new Map([
    ['{', '}'],
    ['(', ')'],
    ['[', ']']
  ])

  const stack = [] as string[]

  for (let el of [...s]) {
    if (map.has(el)) {
      stack.push(el)
    } else {
      if (map.get(stack.pop()) !== el) return false
    }
  }

  return stack.length === 0
}

File: solved/go/LC-0020-ValidParentheses/isValid.go
// Package isvalid solves X problem from Y
// URL: https://leetcode.com/problems/valid-parentheses
package isvalid

// isValid is the primary function for the problem
func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}

	st := []rune{}
	open := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}

	for _, r := range s {
		if pair, ok := open[r]; ok {
			st = append(st, pair)
			continue
		}

		l := len(st) - 1
		if l < 0 || r != st[l] {
			return false
		}
		st = st[:l]
	}

	return len(st) == 0
}
